lib/taskana-test-api/src/test/java/pro/taskana/testapi/util/ServiceProviderExtractorTest.java

```diff
@@ -0,0 +1,341 @@
+package pro.taskana.testapi.util;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatCode;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static pro.taskana.testapi.util.ServiceProviderExtractor.extractServiceProviders;
+
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalInt;
+import org.assertj.core.api.ThrowableAssert.ThrowingCallable;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.TestInstance.Lifecycle;
+import org.junit.platform.commons.JUnitException;
+
+import pro.taskana.spi.priority.api.PriorityServiceProvider;
+import pro.taskana.spi.task.api.CreateTaskPreprocessor;
+import pro.taskana.task.api.models.Task;
+import pro.taskana.task.api.models.TaskSummary;
+import pro.taskana.testapi.WithServiceProvider;
+
+class ServiceProviderExtractorTest {
+
+  static class StaticCreateTaskPreprocessor implements CreateTaskPreprocessor {
+    @Override
+    public void processTaskBeforeCreation(Task taskToProcess) {
+      // implementation not important for the tests
+    }
+  }
+
+  static class DummyTaskPreprocessor1 implements CreateTaskPreprocessor {
+    @Override
+    public void processTaskBeforeCreation(Task taskToProcess) {
+      // implementation not important for the tests
+    }
+  }
+
+  static class DummyTaskPreprocessor2 implements CreateTaskPreprocessor {
+    @Override
+    public void processTaskBeforeCreation(Task taskToProcess) {
+      // implementation not important for the tests
+    }
+  }
+
+  static class DummyPriorityServiceProvider1 implements PriorityServiceProvider {
+    @Override
+    public OptionalInt calculatePriority(TaskSummary taskSummary) {
+      // implementation not important for the tests
+      return OptionalInt.empty();
+    }
+  }
+
+  static class DummyPriorityServiceProvider2 implements PriorityServiceProvider {
+    @Override
+    public OptionalInt calculatePriority(TaskSummary taskSummary) {
+      // implementation not important for the tests
+      return OptionalInt.empty();
+    }
+  }
+
+  private static class PrivateStaticCreateTaskPreprocessor implements CreateTaskPreprocessor {
+    @Override
+    public void processTaskBeforeCreation(Task taskToProcess) {
+      // implementation not important for the tests
+    }
+  }
+
+  @Nested
+  @TestInstance(Lifecycle.PER_CLASS)
+  class ServiceProviderInstantiation {
+
+    @Test
+    void should_InstantiateServiceProvider_When_ServiceProviderIsTopLevelClass() {
+      @WithServiceProvider(
+          serviceProviderInterface = CreateTaskPreprocessor.class,
+          serviceProviders = TopLevelCreateTaskPreprocessor.class)
+      class Clazz {}
```

Comment:
The name "Clazz" sounds unclear. Maybe we could rename it to "ExampleTestClass", "ExampleTestClassUsingServiceProviders" or something else?